docs(gatsby-plugin-image): Update readme (#29076)
gatsbyjs/gatsby: packages/gatsby-plugin-image/README.md

# gatsby-plugin-image (beta)

_The new Gatsby Image plugin is currently in beta, but you can try it out now and see what it can do for the performance of your site._

Adding responsive images to your site while maintaining high performance scores can be difficult to do manually. The Gatsby Image plugin handles the hard parts of producing images in multiple sizes and formats for you!

For full documentation on all configuration options, see [the reference guide](https://www.gatsbyjs.com/docs/reference/)

## Contents

- [Installation](#installation)
- [Using the Gatsby Image components](#using-the-gatsby-image-components)
  - [Static images](#static-images)
  - [Dynamic images](#dynamic-images)
- [Migrating to gatsby-plugin-image](#migrating)

## Installation

> 1. Install `gatsby-plugin-image`, `gatsby-source-filesystem` and `gatsby-plugin-sharp`:

Reviewer: Should we have a switch case where source-filesystem is for static and transformer is for gatsbyImage? You won't always need it, right?

Author: No, not always it's true, but it's not as simple as 1:1 though

Adding responsive images to your site while maintaining high performance scores can be difficult to do manually. The Gatsby Image plugin handles the hard parts of producing images in multiple sizes and formats for you!

For full documentation on all configuration options, see [the Gatsby Image Plugin reference guide](https://www.gatsbyjs.com/docs/reference/built-in-components/gatsby-plugin-image/)

## Contents

- [Installation](#installation)
- [Using the Gatsby Image components](#using-the-gatsby-image-components)
  - [Static images](#static-images)
  - [Dynamic images](#dynamic-images)
- [Migrating to gatsby-plugin-image](#migrating)

## Installation

1. Install `gatsby-plugin-image` and `gatsby-plugin-sharp`. Additionally install `gatsby-source-filesystem` if you are using static images, and `gatsby-transformer-sharp` if you are using dynamic images.

```shell
npm install gatsby-plugin-image gatsby-plugin-sharp gatsby-source-filesystem gatsby-transformer-sharp
```

1. Upgrade `gatsby` to at least `2.24.78`.

2. Add the plugins to your `gatsby-config.js`:

```javascript
module.exports = {
  plugins: [
    `gatsby-plugin-image`,
    `gatsby-plugin-sharp`,

>     `gatsby-source-filesystem`, // Needed for static images

Reviewer 2: Defining `gatsby-source-filesystem` like this will thrown an error:

![image](https://user-images.githubusercontent.com/16143594/105854158-d3797000-5fe6-11eb-864c-576453a4621d.png)

Reviewer: Missed that. Also, they don't need to configure that one. Only need to install it.

module.exports = {
  plugins: [
    `gatsby-plugin-image`,

>     `gatsby-psource-filesystem`,

Reviewer: ```suggestion
    `gatsby-source-filesystem`,
```

    `gatsby-plugin-sharp`,
    // `gatsby-transformer-sharp` // Needed for dynamic images
  ],
}
```

## Using the Gatsby Image components

> ### Decide which component to use

Reviewer: ```suggestion
### Deciding which component to use
```
We say using above, so for consistencies sake

> The Gatsby Image plugin includes two image components: one for static and one for dynamic images. The simplest way to decide which you need to is to ask yourself: _"will this image be the same every time the component or template is used?"_. If it will always be the same, then use `StaticImage`. If it will change, whether through data coming from a CMS or different values passed to a component each time you use it, then it is a dynamic image and you should use the `GatsbyImage` component.

Reviewer: ```suggestion
The Gatsby Image plugin includes two image components: one for static and one for dynamic images. An effective way to decide which you need to is to ask yourself: _"will this image be the same every time the component or template is used?"_. If it will always be the same, then use `StaticImage`. If it will change, whether through data coming from a CMS or different values passed to a component each time you use it, then it is a dynamic image and you should use the `GatsbyImage` component.
```

### Static images

> If you are using an image that will be the same each time the component is used, such as a logo or front page hero image, you can use the `StaticImage` component. The image can be either a local file in your project, or an image hosted on a remote server. Any remote images are downloaded and resized at build time.

Reviewer: ```suggestion
If you are using an image that will be the same each time the component is used, such as a logo or front page hero image, you can use the `StaticImage` component. The image can be a local file in your project or an image hosted on a remote server. Any remote images are downloaded and resized at build time.
```

Reviewer: Should we say URL explicitly?

1. **Add the image to your project.**

>    If you are using a local image, copy it into the project. A folder such as `src/images` is a good choice.

Reviewer: Should we mention anything about source-fileystem here?

Author: You don't need source-filesystem for this case. It is, counter-intuitively, only needed for remote images (because we need createRemoteFileNode)

Reviewer: We import `gatsby-source-filesystem/create-file-node` for the createFileNode case.

   export function Kitten() {
     return <StaticImage src="https://placekitten.com/800/600" alt="A kitten" />
   }
   ```

   When you build your site, the `StaticImage` component will load the image from your filesystem or from the remote URL, and it will generate all the sizes and formats that you need to support a responsive image.

   Because the image is loaded at build time, you cannot pass the filename in as a prop, or otherwise generate it outside of the component. It should either be a static string, or a local variable in the component's scope.

   **Important:** Remote images are downloaded and resized at build time. If the image is changed on the other server, it will not be updated on your site until you rebuild.

3. **Configure the image.**

   You configure the image by passing props to the `<StaticImage />` component. You can change the size and layout, as well as settings such as the type of placeholder used when lazy loading. There are also advanced image processing options available. You can find the full list of options [in the API docs](https://www.gatsbyjs.com/docs/reference/built-in-components/gatsby-plugin-image).

>    This component renders a 200px by 200px image of a dinosaur. Before loading it will have a blurred, low-resolution placeholder. It uses the `"fixed"` layout, which means the image does not resize with its container.

Reviewer: I might move this underneath the example. Because we import Dino in the example two above, I expected it was talking about the defaults applied there.

   ```jsx
   import { StaticImage } from "gatsby-plugin-image"

   export function Dino() {
     return (
       <StaticImage
         src="../images/dino.png"
         alt="A dinosaur"
         placeholder="blurred"
         layout="fixed"
         width={200}
         height={200}
       />
     )
   }
   ```

#### Restrictions on using `StaticImage`

The images are loaded and processed at build time, so there are restrictions on how you pass props to the component. The values need to be statically-analyzed at build time, which means you can't pass them as props from outside the component, or use the results of function calls, for example. You can either use static values, or variables within the component's local scope. See the following examples:

This does not work:

```js
// ⚠️ Doesn't work

export function Logo({ logo }) {
  // You can't use a prop passed into the parent component
  return <StaticImage src={logo}>
}
```

...and nor does this:

> ```js

Reviewer: ```suggestion
```jsx
```

// ⚠️ Doesn't work

export function Dino() {
    // Props can't come from function calls
    const width = getTheWidthFromSomewhere();
    return <StaticImage src="trex.png" width={width}>
}
```

You can use variables and expressions if they're in the scope of the file, e.g.:

> ```js

Reviewer: ```suggestion
```jsx
```

// OK
export function Dino()  {
    // Local variables are fine
    const width = 300
    return <StaticImage src="trex.png" width={width}>
}
```

> ```js

Reviewer: ```suggestion
```jsx
```

// Also OK

// A variable in the same file is fine.
const width = 300

export function Dino()  {
    // This works because the value can be statically-analyzed
    const height = width * 16 / 9
    return <StaticImage src="trex.png" width={width} height={height}>
}
```

If you find yourself wishing you could use a prop for the image `src` then it's likely that you should be using a dynamic image.

### Dynamic images

If you need to have dynamic images (such as if they are coming from a CMS), you can load them via GraphQL and display them using the `GatsbyImage` component.

> 1. **Add the image to your page query.**

Reviewer: Might be worth mentioning gatsby-transformer-sharp again here

   Any GraphQL File object that includes an image will have a `childImageSharp` field that you can use to query the image data. The exact data structure will vary according to your data source, but the syntax is like this:

   ```graphql
   query {
     blogPost(id: { eq: $Id }) {
       title
       body
       avatar {
         childImageSharp {
           gatsbyImageData(width: 200)
         }
       }
     }
   }
   ```

2. **Configure your image.**

>    You configure the image by passing arguments to the `gatsbyImageData` resolver. You can change the size and layout, as well as settings such as the type of placeholder used when lazy loading. There are also advanced image processing options available. You can find the full list of options in the API docs.

Reviewer: Link to the API docs again in case people are jumping to this section?

   ```graphql
   query {
     blogPost(id: { eq: $Id }) {
       title
       body
       author
       avatar {
         childImageSharp {
           gatsbyImageData(

>              width: 200

Reviewer: ```suggestion
             width: 200,
```

Author: Commas are optional when it's multiline. Prettier removes them.

Reviewer: I legit just realized this when I was doing it on my site. Strange.

Author: I accepted all of the suggesitons, but then started to edit locally before committing then realised prettier just removed them again!

Reviewer: Ya, I think just leave them out. Hopefully people won't have copy/paste issues. But if it's valid syntax they shouldn't.

>              placeholder: BLURRED

Reviewer: ```suggestion
             placeholder: BLURRED,
```

             formats: [AUTO, WEBP, AVIF]
           )
         }
       }
     }
   }
   ```

3. **Display the image.**

   You can then use the `GatbsyImage` component to display the image on the page. The `getImage()` function is an optional helper to make your code easier to read. It takes a `File` and returns `file.childImageSharp.gatsbyImageData`, which can be passed to the `GatsbyImage` component.

   ```jsx
   import { graphql } from "gatsby"
   import { GatsbyImage, getImage } from "gatsby-plugin-image"

   function BlogPost({ data }) {
     const image = getImage(data.blogPost.avatar)
     return (
       <section>
         <h2>{data.blogPost.title}</h2>
         <GatsbyImage image={image} alt={data.blogPost.author} />
         <p>{data.blogPost.body}</p>
       </section>
     )
   }

   export const pageQuery = graphql`
     query {
       blogPost(id: { eq: $Id }) {
         title
         body
         author
         avatar {
           childImageSharp {
             gatsbyImageData(

>                width: 200

Reviewer: ```suggestion
               width: 200,
```

>                placeholder: BLURRED

Reviewer: ```suggestion
               placeholder: BLURRED,
```
